src/tui/mod.rs

```diff
@@ -15,24 +16,196 @@ use ratatui::prelude::Rect;
+        &mut self.state
+    }
+}
+
+impl Tui {
+    fn new(mut terminal: TerminalGuard) -> Self {
+        let area = terminal.get_frame().size();
+        Self {
+            terminal,
+            size: area,
+            state: Default::default(),
+        }
+    }
+
+    fn half_height(&mut self) -> usize {
+        (self.size.height / 2) as usize
+    }
+
+    fn scroll_up(&mut self, amount: usize) {
+        self.scroll_offset = self.scroll_offset.saturating_sub(amount);
+    }
+
+    fn scroll_down(&mut self, amount: usize) {
+        self.scroll_offset = min(self.scroll_max(), self.scroll_offset.saturating_add(amount));
+    }
+
+    fn scroll_max(&mut self) -> usize {
+        self.line_count.saturating_sub(self.half_height())
+    }
+
+    fn scroll_to(&mut self, scroll_offset: usize) {
+        self.scroll_offset = min(self.scroll_max(), scroll_offset);
+    }
+
+    fn push_line(&mut self, line: String) {
+        self.scrollback.extend(line.into_bytes());
+        self.scrollback.push(b'\n');
+        self.line_count += 1;
+    }
+
+    #[instrument(level = "trace", skip(self))]
+    fn render(&mut self) -> miette::Result<()> {
+        let mut render_result = Ok(());
+        self.terminal
+            .draw(|frame| {
+                self.size = frame.size();
+                let buffer = frame.buffer_mut();
+                render_result = self.state.render_inner(self.size, buffer);
+            })
+            .into_diagnostic()
+            .wrap_err("Failed to draw to terminal")?;
+
+        Ok(())
+    }
+
+    #[instrument(level = "trace", skip(self))]
+    fn handle_event(&mut self, event: Event) -> miette::Result<()> {
+        // TODO: Steal Evan's declarative key matching macros?
+        // https://github.com/evanrelf/indigo/blob/7a5e8e47291585cae03cdf5a7c47ad3bcd8db3e6/crates/indigo-tui/src/key/macros.rs
+        match event {
+            Event::Mouse(mouse) if mouse.kind == MouseEventKind::ScrollUp => {
+                self.scroll_up(SCROLL_AMOUNT);
+            }
+            Event::Mouse(mouse) if mouse.kind == MouseEventKind::ScrollDown => {
+                self.scroll_down(SCROLL_AMOUNT);
+            }
+            Event::Key(key) => match key.modifiers {
+                KeyModifiers::NONE => match key.code {
+                    KeyCode::Char('j') => {
+                        self.scroll_down(1);
+                    }
+                    KeyCode::Char('k') => {
+                        self.scroll_up(1);
+                    }
+                    KeyCode::Char('g') => {
+                        self.scroll_to(0);
+                    }
+                    _ => {}
+                },
+
+                #[allow(clippy::single_match)]
+                KeyModifiers::SHIFT => match key.code {
+                    KeyCode::Char('G') => {
+                        self.scroll_to(usize::MAX);
+                    }
+                    _ => {}
+                },
```

Comment:
I wonder if there are terminals that report this as `SHIFT` + `g` and not `SHIFT` + `G`? Maybe writing these `SHIFT`-based keys like this would be more robust:

```suggestion
                KeyModifiers::SHIFT => match key.code {
                    KeyCode::Char('g' | 'G') => {
                        self.scroll_to(usize::MAX);
                    }
                    _ => {}
                },
```